src/ChluIPFS.js

```diff
@@ -0,0 +1,184 @@
+            ]
+        }
+    }
+};
+
+class ChluIPFS {
+
+    constructor(options = {}){
+        const additionalOptions = {
+            // TODO: review this. We should have a persistent store in node. Volatile is fine in the browser
+            store: String(Math.random() + Date.now())
+        };
+        // TODO: review persisting data in orbitDb. we don't need persistence in the browser but it would help.
+        this.orbitDbDirectory = options.orbitDbDirectory || '../orbit-db';
+        this.ipfsOptions = Object.assign(
+            {},
+            defaultIPFSOptions,
+            additionalOptions,
+            options.ipfs || {}
+        );
+        this.type = options.type;
+        if (Object.values(constants.types).indexOf(this.type) < 0) {
+            throw new Error('Invalid type');
+        }
+        this.utils = ipfsUtils;
+        this.events = new EventEmitter();
+    }
+    
+    async start(){
+        if (!this.ipfs) {
+            this.ipfs = await this.utils.createIPFS(this.ipfsOptions );
+        }
+        // OrbitDB setup
+        if (!this.orbitDb) {
+            this.orbitDb = new OrbitDB(this.ipfs, this.orbitDbDirectory);
+        }
+        if (this.type === constants.types.customer && !this.db) {
+            this.db = await this.orbitDb.feed('chlu-experimental-customer-review-updates');
+        }
+        // PubSub setup
+        if (!this.room) {
+            this.room = Room(this.ipfs, 'chlu-experimental');
+            // handle events
+            this.room.on('message', message => this.handleMessage(message));
+            // wait for room subscription
+            await new Promise(resolve => {
+                this.room.on('subscribed', () => resolve());
+            });
+        }
+        // If customer, also wait for at least one peer to join the room (TODO: review this)
+        if (this.type === constants.types.customer) {
+            if (this.room.getPeers().length === 0) {
+                await new Promise(resolve => {
+                    this.room.on('peer joined', () => resolve());
+                });
+            }
+        } else if (this.type === constants.types.service) {
+            this.startServiceNode();
+        }
+        return true;
+    }
+
+    async stop() {
+        // leaving the pubsub room is already handled by the library on IPFS stop
+        if (this.db) await this.db.close();
+        await this.ipfs.stop();
+        this.db = undefined;
+        this.room = undefined;
+        this.orbitDb = undefined;
+        this.ipfs = undefined;
+    }
+
+    async pin(multihash){
+        // broadcast start of pin process
+        await this.room.broadcast(this.utils.encodeMessage({ type: constants.eventTypes.pinning, multihash }));
+        if (this.ipfs.pin) {
+            await this.ipfs.pin.add(multihash, { recursive: true });
+        } else {
+            // TODO: Chlu service node need to be able to pin, so we should support using go-ipfs
```

Comment:
I do think the chlu service node should really be based on go-ipfs. Couple of reasons, 1) probably more mature implementation with more features and 2) probably less resource intensive. For the same reason I am in favour of a chlu-service-node-js and later a chlu-service-node-go repo

Reply:
I agree, when possible we should use go-ipfs. Lucky for us, we can talk to go-ipfs using js-ipfs-api, which has the same exact calls to js-ipfs with the difference that js-ipfs-api forwards calls to go-ipfs via HTTP and js-ipfs executes them in javascript.

This means we can support running ChluIPFS on top of go-ipfs (or any other ipfs implementation really) with configuration options.

We need to verify that go-ipfs can connect to js-ipfs nodes running in the browser. This might be an issue, in the browser we can only use the websocket-star transport and it's unclear wether it can connect to the go-ipfs websocket transport at the moment (https://github.com/libp2p/go-ws-transport) even though it should eventually.

I created an issue for this: https://github.com/ChluNetwork/chlu-ipfs-support/issues/3